api/specs/web-server/_common.py

```diff
@@ -5,64 +5,74 @@
 from collections.abc import Callable
 from pathlib import Path
-from typing import Any, ClassVar, NamedTuple
+from typing import Annotated, NamedTuple, Optional, Union, get_args, get_origin
 
-import yaml
 from common_library.json_serialization import json_dumps
 from common_library.pydantic_fields_extension import get_type
-from fastapi import FastAPI, Query
+from fastapi import Query
 from models_library.basic_types import LogLevel
-from pydantic import BaseModel, Field, create_model
+from pydantic import BaseModel, ConfigDict, Field, Json, create_model
 from pydantic.fields import FieldInfo
-from servicelib.fastapi.openapi import override_fastapi_openapi_method
 
 CURRENT_DIR = Path(sys.argv[0] if __name__ == "__main__" else __file__).resolve().parent
 
 
-def _create_json_type(**schema_extras):
-    class _Json(str):
-        __slots__ = ()
+def _replace_basemodel_in_annotation(annotation, new_type):
+    origin = get_origin(annotation)
 
-        @classmethod
-        def __modify_schema__(cls, field_schema: dict[str, Any]) -> None:
-            # openapi.json schema is corrected here
-            field_schema.update(
-                type="string",
-                # format="json-string" NOTE: we need to get rid of openapi-core in web-server before using this!
-            )
-            if schema_extras:
-                field_schema.update(schema_extras)
+    # Handle Annotated
+    if origin is Annotated:
+        args = get_args(annotation)
+        base_type = args[0]
+        metadata = args[1:]
+        if isinstance(base_type, type) and issubclass(base_type, BaseModel):
+            # Replace the BaseModel subclass
+            base_type = new_type
 
-    return _Json
+        return Annotated[(base_type, *metadata)]
+
+    # Handle Optionals, Unions, or other generic types
+    if origin in (Optional, Union, list, dict, tuple):  # Extendable for other generics
+        new_args = tuple(
+            _replace_basemodel_in_annotation(arg, new_type)
+            for arg in get_args(annotation)
+        )
+        return origin[new_args]
+
+    # Replace BaseModel subclass directly
+    if isinstance(annotation, type) and issubclass(annotation, BaseModel):
+        return new_type
+
+    # Return as-is if no changes
+    return annotation
 
 
 def as_query(model_class: type[BaseModel]) -> type[BaseModel]:
     fields = {}
     for field_name, field_info in model_class.model_fields.items():
 
-        field_type = get_type(field_info)
-        default_value = field_info.default
-
-        kwargs = {
+        field_default = field_info.default
+        assert not field_info.default_factory  # nosec
+        query_kwargs = {
             "alias": field_info.alias,
             "title": field_info.title,
             "description": field_info.description,
             "metadata": field_info.metadata,
-            "json_schema_extra": field_info.json_schema_extra,
+            "json_schema_extra": field_info.json_schema_extra or {},
```

Comment:
Q: why the default to {}? I thought json_schema_extra can have multiple types and also None?

Reply:
Yes, it can be None. I found it when the script raise an exception :-)

Reply:
changed